modules/tensor_mechanics/src/userobjects/TensorMechanicsPlasticMohrCoulombMulti.C

```diff
@@ -252,173 +351,547 @@ TensorMechanicsPlasticMohrCoulombMulti::activeConstraints(const std::vector<Real
+  mooseAssert(yf.size() == 6, "TensorMechanicsPlasticMohrCoulombMulti: Custom tip-return algorithm must be supplied with yf of size 6, whereas yours is " << yf.size());
+
+  dpm[3] = initial_guess;
+  sinphi = std::sin(phi(intnl_old + dpm[3]));
+  Real cosphi = std::cos(phi(intnl_old + dpm[3]));
+  Real coh = cohesion(intnl_old + dpm[3]);
+  cohcos = coh*cosphi;
+
+  Real alpha = n[3][0] - n[3][2] - (n[3][2] + n[3][0])*sinphi;
+  Real deriv_phi;
+  Real dalpha;
+  Real beta = eigvals[2] + eigvals[0];
+  Real residual = alpha*dpm[3] + eigvals[2] - eigvals[0] + beta*sinphi - 2*cohcos; // this is 2*yf[3]
+  Real deriv_coh;
+  Real jacobian;
+  unsigned int iter = 0;
+  do {
+    deriv_phi = dphi(intnl_old + dpm[3]);
+    dalpha = -(n[3][2] + n[3][0])*cosphi*deriv_phi;
+    deriv_coh = dcohesion(intnl_old + dpm[3]);
+    jacobian = alpha + dalpha*dpm[3] + beta*cosphi*deriv_phi - 2*deriv_coh*cosphi + 2*coh*sinphi*deriv_phi;
+
+    dpm[3] -= residual/jacobian;
+    if (iter > _max_iters) // not converging
+    {
+      nr_converged = false;
+      return false;
+    }
+
+    sinphi = std::sin(phi(intnl_old + dpm[3]));
+    cosphi = std::cos(phi(intnl_old + dpm[3]));
+    coh = cohesion(intnl_old + dpm[3]);
+    cohcos = coh*cosphi;
+    alpha = n[3][0] - n[3][2] - (n[3][2] + n[3][0])*sinphi;
+    residual = alpha*dpm[3] + eigvals[2] - eigvals[0] + beta*sinphi - 2*cohcos;
+    iter ++;
+  } while (residual*residual > _f_tol*_f_tol);
+
+  // so the NR process converged, but we must
+  // check Kuhn-Tucker
+  nr_converged = true;
+  if (dpm[3] < -ep_plastic_tolerance)
+    // Kuhn-Tucker failure
+    return false;
+
+  for (unsigned i = 0 ; i < 3 ; ++i)
+    returned_stress(i, i) = eigvals[i] - dpm[3]*n[3][i];
+  yieldFunctionEigvals(returned_stress(0, 0), returned_stress(1, 1), returned_stress(2, 2), sinphi, cohcos, yf);
+
+  // by construction abs(yf[3]) = abs(residual/2) < _f_tol/2
+  if (yf[0] > _f_tol || yf[1] > _f_tol || yf[2] > _f_tol || yf[4] > _f_tol || yf[5] > _f_tol)
+    // Kuhn-Tucker failure
+    return false;
+
+  // success!
+  dpm[0] = dpm[1] = dpm[2] = dpm[4] = dpm[5] = 0;
+  return true;
 }
 
-Real
-TensorMechanicsPlasticMohrCoulombMulti::dcohesion(const Real internal_param) const
-{
-  return _cohesion.derivative(internal_param);
-}
 
-Real
-TensorMechanicsPlasticMohrCoulombMulti::phi(const Real internal_param) const
-{
-  return _phi.value(internal_param);
-}
 
-Real
-TensorMechanicsPlasticMohrCoulombMulti::dphi(const Real internal_param) const
+bool
+TensorMechanicsPlasticMohrCoulombMulti::returnEdge000101(const std::vector<Real> & eigvals, const std::vector<std::vector<Real> > & n, std::vector<Real> & dpm, RankTwoTensor & returned_stress, const Real & intnl_old, Real & sinphi, Real & cohcos, const Real & initial_guess, const Real & mag_E, bool & nr_converged, const Real & ep_plastic_tolerance, std::vector<Real> & yf) const
```

Comment:
Heheheh, i love that prophet.  Here, and elsewhere, the "000101" etc, tells the reader which of the 6 surfaces of the MohrCoulomb yield criterion are active.  It's very important to know this otherwise the return* functions don't make sense.  I need to highlight this in the in-code documentation.